pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -703,17 +703,24 @@ func (m *kubeGenericRuntimeManager) purgeInitContainers(pod *v1.Pod, podStatus *
 // index of next init container to start, or done if there are no further init containers.
 // Status is only returned if an init container is failed, in which case next will
 // point to the current container.
-func findNextInitContainerToRun(pod *v1.Pod, podStatus *kubecontainer.PodStatus) (status *kubecontainer.ContainerStatus, next *v1.Container, done bool) {
+func findNextInitContainerToRun(pod *v1.Pod, podStatus *kubecontainer.PodStatus) (status *kubecontainer.ContainerStatus, next *v1.Container, done bool, changed bool) {
 	if len(pod.Spec.InitContainers) == 0 {
-		return nil, nil, true
+		return nil, nil, true, false
 	}
 
 	// If there are failed containers, return the status of the last failed one.
 	for i := len(pod.Spec.InitContainers) - 1; i >= 0; i-- {
 		container := &pod.Spec.InitContainers[i]
 		status := podStatus.FindContainerStatusByName(container.Name)
-		if status != nil && isInitContainerFailed(status) {
-			return status, container, false
+		if status != nil {
+			expectedHash, actualHash, changed := containerChanged(container, status)
+			if changed {
+				klog.V(4).Infof("Init container spec hash changed (%d vs %d).", actualHash, expectedHash)
+				return nil, nil, false, true
```

Comment:
should this logic live inside findNextInitContainerToRun  ?

Reply:
leaving this logic here is to re-use the `for` loop inside findNextInitContainerToRun. I suppose that `findNextInitContainerToRun` is to compute next actions for init containers. So i think it is fine that we have hash checking here.